### PR TITLE
Move ``get_world_size_and_rank`` to utils

### LINKED ISSUE #2151: Make distributed available outside of torchtune.training
Modules in torchtune.data can't import get_world_size_and_rank from torchtune.training._distributed because of circular imports. Making this available in torchtune._distributed instead would fix this issue
https://github.com/pytorch/torchtune/pull/1929#discussion_r1879199687_

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Closes #2151 

#### Changelog
What are the changes made in this PR?
* Move ``get_world_size_and_rank`` from torchtune/training to torchtune/utils. **Why does this make sense?** B/c it's more of a utility anyways and can be used outside of the training loop
* Update all tests and call sites accordingly.
* Deprecate old ``get_world_size_and_rank``

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/utils/_device.py [hunk 1]
```diff
@@ -21,6 +21,19 @@
     BlockMask = torch.Tensor
 
 
+def get_world_size_and_rank() -> Tuple[int, int]:
```
Comment (ebsmothers) [inline torchtune/utils/_device.py:24]: I understand the goal of this and agree with moving this function down into utils. But I also wonder whether device.py is really the best place for it (I know we already have `_get_local_rank` in here, and I know that having separate `utils/distributed.py` and `training/distributed.py` is probably confusing)

Comment (joecummings) [reply torchtune/utils/_device.py:24]: I'm in favor of keeping it in here.

**Review by ebsmothers — APPROVED** [review]
